### PR TITLE
IGNITE-17535 Implementing a hash index B+Tree

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-17535

### CODE REVIEW

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/hash/RemoveHashIndexRowInvokeClosure.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    public OperationType operationType() {
+        return operationType;
+    }
+
+    /**
+     * Method to call after {@link BplusTree#invoke(Object, Object, InvokeClosure)} has completed.
+     *
+     * @throws IgniteInternalCheckedException If failed to remove data from the free list.
+     */
+    public void afterCompletion() throws IgniteInternalCheckedException {
```
Comment (SammyVimes) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/hash/RemoveHashIndexRowInvokeClosure.java:88]: This method is unused, is it on purpose?

**Review by SammyVimes — APPROVED** [review]
LGTM!

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/freelist/IndexColumns.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        assert valueBuffer != null;
+
+        return valueBuffer.limit();
+    }
+
+    /**
+     * Returns a byte buffer that contains binary tuple data.
+     */
+    public ByteBuffer valueBuffer() {
+        return valueBuffer;
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/freelist/IndexColumns.java:87]: Hell no! Array access doesn't work for read-only buffers, making unsafe access almost impossible.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/hash/RemoveHashIndexRowInvokeClosure.java (hunk 1 above)
Comment (ibessonov) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/hash/RemoveHashIndexRowInvokeClosure.java:88]: Yes, class is unused as well. For some reason, such method is missing in the closure interface, which is kinda stupid and _maybe_ we should fix that later

**Review by SammyVimes — APPROVED** [review]
LGTM

**Pull request merged**